SimPEG/regularization/coupling.py

```diff
@@ -0,0 +1,744 @@
+
+            return x_grad, y_grad, z_grad
+
+    def gradient_list(self, x_grad, y_grad, z_grad=np.array([])):
+        '''
+        Concatenates gradient components into a single array.
+        [[x_grad1, y_grad1, z_grad1],
+         [x_grad2, y_grad2, z_grad2],
+         [x_grad3, y_grad3, z_grad3],...]
+
+        :param numpy.ndarray x_grad: x-gradients
+        :param numpy.ndarray y_grad: y-gradients
+        :param numpy.ndarray z_grad: z-gradients
+
+        :rtype: numpy.ndarray
+        :return: gradient_vector: array where each row represents a model cell,
+                 and each column represents a component of the gradient.
+        '''
+        if z_grad.size == 0:
+            gradient_vector = np.c_[x_grad, y_grad]
+        else:
+            gradient_vector = np.c_[x_grad, y_grad, z_grad]
+
+        return gradient_vector
+
+
+    def gradient_applitude_inv(self, m1, m2, fltr=True, fltr_per=0.05):
+        '''
+        Computes the norms of the gradients for two models.
+
+        :param numpy.ndarray m1: model1
+        :param numpy.ndarray m2: model2
+        :param bool reduced_space: if True, return only norms of active cells.
+        :param bool fltr: if True, filter out predefined percentage of lowest norms.
+        :param float fltr_per: Percentage of lowest norms to be filtered out.
+                               Default is 0.05
+
+        :rtype tuple of numpy.ndarray
+        :return (norms_1, norms_2, tot_norms)
+                norms_1: np.array containing reciprocals of the gradient norms for model1
+                norms_2: np.array containing reciprocals of the gradient norms for model2
+                tot_norms: np.array containing element-wise product of norms_1 and norms_2.
+        '''
+        per = int(fltr_per*self.regmesh.nC)
+
+        if self.regmesh.mesh.dim == 2:
+            Dx_m1, Dy_m1 = self.calculate_gradient(m1)
+            Dx_m2, Dy_m2 = self.calculate_gradient(m2)
+
+            norms_1 = np.sqrt(Dx_m1**2 + Dy_m1**2)
+            # compute (1 / applitude of gradients) and background when norms < 1e-10
+            norms_1 = np.divide(1, norms_1, out=np.zeros_like(norms_1), where=norms_1>1e-10)
+            norms_2 = np.sqrt(Dx_m2**2 + Dy_m2**2)
+            norms_2 = np.divide(1, norms_2, out=np.zeros_like(norms_2), where=norms_2>1e-10)
+
+        elif self.regmesh.mesh.dim == 3:
+            Dx_m1, Dy_m1, Dz_m1 = self.calculate_gradient(m1)
+            Dx_m2, Dy_m2, Dz_m2 = self.calculate_gradient(m2)
+
+            norms_1 = np.sqrt(Dx_m1**2 + Dy_m1**2 + Dz_m1**2)
+            norms_1 = np.divide(1, norms_1, out=np.zeros_like(norms_1), where=norms_1>1e-10)
+            norms_2 = np.sqrt(Dx_m2**2 + Dy_m2**2 + Dz_m2**2)
+            norms_2 = np.divide(1, norms_2, out=np.zeros_like(norms_2), where=norms_2>1e-10)
+
+        # set lowest 5% of norms (largest 5% of 1/norms) to 0.0
+        if fltr:
+            inds1 = norms_1.argsort()[-per:]
```

Comment:
@kimjaed Hi, Jae, it's great to have you here as the first developer of this PR! I kept another filter to avoid extremely small amplitude of gradient. But I removed the filter "set lowest 5% of norms (largest 5% of 1/norms) to 0.0" shown line 211 - 214 above, because this filter is not a common case and the default is False in code as well. What do you think about it?